### ISSUE TITLE
Can't type mixins anymore because "self must be a supertype of its class"

### ISSUE THREAD

**Issue by vberlier** [microsoft/pyright#1716]
This is kind of a contrived example but this doesn't work anymore in the latest version:

```python
class FooMixin:
    # Type of parameter "self" must be a supertype of its class "FooMixin"
    def foo(self: Callable[[], Any]):
        self()
        self()

class Bar(FooMixin):
    def __call__(self):
        print("bar")

Bar().foo()
```

However you can make it work with a custom protocol:

```python
class SupportsCall(Protocol):
    def __call__(self) -> Any: ...

class FooMixin:
    def foo(self: SupportsCall):  # No errors
        self()
        self()
```

It would be nice to at least allow types from the typing module.

**Comment by erictraut** [comment]
The usage of mixins in your example is not type safe. If you add that mixin to a class that doesn't support a `__call__` method, it won't work. The type-safe way to implement this is to use a protocol, as you've done in the second example.

Mypy also emits an error for your first code sample.

**Comment by vberlier** [comment]
Good point, I didn't check mypy. But still it's kind of weird because there's no difference between `Callable` and the protocol when the class doesn't implement `__call__`.

```python
class Bar(FooMixin):
    pass

# Could not bind method "foo" because "Bar" is not assignable to parameter "self"
Bar().foo()
```

In both cases the error is reported when accessing the method, as expected. I don't see how using the protocol makes it more type-safe because there's no difference from the user's perspective. What exactly doesn't "work" if I use `Callable` and the class doesn't implement `__call__`?

Fwiw, I have several dict-like objects and I use mixins to compose methods into the different classes. For example I have a `MergeMixin` that does some recursive merging:

```python
class MergeMixin:
    def merge(self: MutableMapping[K, V], other: Mapping[K, V]): ...

class ThingContainer(MergeMixin, MutableMapping[str, Thing]): ...
```

Should I actually define all the methods implemented in `MutableMapping` in a custom protocol?

**Comment by erictraut** [comment]
The way you're applying the mixin pattern is a bit odd here. When I've used this pattern in the past, I've always tried to avoid the case where the mixin class has knowledge of (or dependencies on) the class that it is being mixed with. 

I understand your use case, but it is fragile. A type checker will not have any way to enforce this constraint, so you're depending on users of the mixin class not to make any mistakes. And if any assumptions change in the future, it will break without detection.

You make a reasonable argument about annotations like `Callable` versus a protocol. The counterargument is that protocols are designed specifically to describe the "shape" of an interface, which is exactly what you want in this use case. That's why both mypy and pyright allow protocols in this particular case but don't allow other nominal (non-protocol) classes to be used in this manner.

Mypy has already set the standard in this case, and I think it's a reasonable standard, so I don't think pyright should deviate from it.

I think your options are to use protocols or use `# type: ignore` to suppress the error.

**Comment by vberlier** [comment]
Thanks for the detailed response. I wasn't really clear but the mixins are actually only used to make things maintainable within the library. Users interact directly with the container classes and don't really need to know that the API for all these containers is factored into mixins. I'm considering getting rid of the mixins in favor of assigning the standalone function in each class that needs it:

```python
def merge_container(a: MutableMapping[K, V], b: Mapping[K, V]): ...

class ThingContainer(MutableMapping[str, Thing]):
    merge = merge_container
    ...
```

However, I'm not sure if this actually recommended either as far as static typing is concerned. Another option would be going through a descriptor:

```python
class ThingContainer(MutableMapping[str, Thing]):
    merge = MergeMethod()
```

But it's really not that different from the mixin relying on the fact that the class implements `MutableMapping`.
